File: src/algorithm/leetcode/Problem31.java
package algorithm.leetcode;

import java.util.Arrays;

public class Problem31 {

    private static void swap(int[] nums, int a, int b) {
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }

    private static void reverse(int[] nums, int start, int end) {
        int i = start, j = end;
        while (j > i) {
            swap(nums, i, j);
            i++;
            j--;
        }
    }

    public static void nextPermutation(int[] nums) {
        int i, j, k;
        for (i = nums.length - 1; i > 0; i--) {
            if (nums[i] > nums[i - 1]) {
                break;
            }
        }
        if (i == 0) {
            Arrays.sort(nums);
            return;
        }
        j = i - 1;
        for (k = nums.length - 1; k >= i; k--) {
            if (nums[k] > nums[j]) {
                break;
            }
        }
        swap(nums, j, k);
        reverse(nums, i, nums.length - 1);
    }

    public static void main(String[] args) {
        int[] nums = {1,3,2,4};
        nextPermutation(nums);
    }

}

File: src/algorithm_cpp/problem31.cpp
#include <iostream>
#include <algorithm>
#include <vector>

//观察排列顺序
//[1,2,3]
//[1,3,2]
//[2,1,3]
//[2,3,1]
//[3,1,2]
//[3,2,1]
//要把尽可能靠后的数变大，还要变得尽可能小

using namespace std;

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int a = -1;
	for (int i = nums.size() - 2; i >= 0; i--) {
	    if (nums[i] < nums[i + 1]) {
		a = i;
		break;
	    }
	}
	if (a == -1) {
	    sort(nums.begin(), nums.end());
	    return;
	}
	int b = nums.size() - 1;
	while (b >= 0 && nums[b] <= nums[a]) {
	    b--;
	}
	swap(nums[a], nums[b]);
	reverse(nums.begin() + a + 1, nums.end());
    }
};

int main() {
    Solution s;
    vector<int> a {1,5,1};
    s.nextPermutation(a);
    for (int item: a) {
	cout << item << ",";
    }
    return 0;
}
